vm/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+//![doc = include_str!("../README.md")]
+
+#![deny(missing_docs)]
+#![deny(clippy::all)]
+#![deny(unused_crate_dependencies)]
+#![deny(unused_extern_crates)]
+
+extern crate alloc;
+
+pub use piecrust::{
+    CallReceipt, CallTree, CallTreeElem, ContractData, Error, PageOpening,
+    Session,
+};
+
+use alloc::vec::Vec;
+use std::fmt::{self, Debug, Formatter};
+use std::path::{Path, PathBuf};
+use std::thread;
+
+use dusk_core::abi::{Metadata, Query};
+use piecrust::{SessionData, VM as PiecrustVM};
+
+use self::host_queries::{
+    host_hash, host_poseidon_hash, host_verify_bls, host_verify_bls_multisig,
+    host_verify_groth16_bn254, host_verify_plonk, host_verify_schnorr,
+};
+
+pub(crate) mod cache;
+pub mod host_queries;
+
+/// Create a new session based on the given `VM`.
+pub fn new_session(
+    vm: &VM,
+    base: [u8; 32],
+    chain_id: u8,
+    block_height: u64,
+) -> Result<Session, Error> {
+    vm.session(
+        SessionData::builder()
+            .base(base)
+            .insert(Metadata::CHAIN_ID, chain_id)?
+            .insert(Metadata::BLOCK_HEIGHT, block_height)?,
+    )
+}
+
+/// Create a new genesis session based on the given [`VM`].
+pub fn new_genesis_session(vm: &VM, chain_id: u8) -> Session {
+    vm.session(
+        SessionData::builder()
+            .insert(Metadata::CHAIN_ID, chain_id)
+            .expect("Inserting chain ID in metadata should succeed")
+            .insert(Metadata::BLOCK_HEIGHT, 0)
+            .expect("Inserting block height in metadata should succeed"),
+    )
+    .expect("Creating a genesis session should always succeed")
+}
```

Comment:
Just a reminder for future refactors

These should be part of `Impl VM` and take `&self`